### PR TITLE
[CDAP-12382] Event payload passing between triggering and triggered program

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-12382

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+package co.cask.cdap.api.schedule;
+
+import java.util.List;
+
+/**
+ * Base class for the composite trigger information to be passed to the triggered program.
+ */
+public abstract class AbstractCompositeTriggerInfo extends TriggerInfo {
+  List<TriggerInfo> unitTriggerInfos;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java]: private final field.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  public AbstractCompositeTriggerInfo(Trigger.Type type, List<TriggerInfo> unitTriggerInfos) {
+    super(type);
+    this.unitTriggerInfos = unitTriggerInfos;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java]: better do a defensive copy to an immutable list.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  /**
+   * @return A list of unique {@link TriggerInfo}'s of the unit triggers in this composite trigger.
+   *         Unit triggers are the triggers which are not composite triggers.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java]: list returned should be immutable (see above and needs documentation).

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ * The program status trigger information to be passed to the triggered program.
+ */
+public class ProgramStatusTriggerInfo extends TriggerInfo {
+
+  private final String namespace;
+  private final ApplicationSpecification applicationSpecification;
+  private final ProgramType programType;
+  private final String program;
+  private final Set<ProgramStatus> triggerStatues;
+  @Nullable
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: Why runId is nullable? Also the type should be `RunId` (same as `RuntimeContext.getRunId()`)

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  private final String runId;
+  @Nullable
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: why this is nullable?

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  private final ProgramStatus programStatus;
+  @Nullable
+  private final WorkflowToken workflowToken;
+  @Nullable
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: This shouldn't be nullable as well.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 4 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: These fields are nullable because of composite trigger. For instance, if the composite trigger is an OR trigger which contains two program status trigger, and only one of them is satisfied with a program run, then the unsatisfied one does not have runId or workflow token

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+   */
+  public String getProgram() {
+    return program;
+  }
+
+  /**
+   * @return All the program statuses that can satisfy the program status trigger.
+   */
+  public Set<ProgramStatus> getTriggerStatues() {
+    return triggerStatues;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: should have a defensive copy to an unmodifiable set (happen in constructor) and returns (see above).

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * the License.
+ */
+
+package co.cask.cdap.api.schedule;
+
+/**
+ * The stream size trigger information to be passed to the triggered program.
+ */
+public class StreamSizeTriggerInfo extends TriggerInfo {
+  String streamNamespace;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java]: private final fields.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  String streamName;
+  int triggerMB;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java]: Similar to dataset partition trigger info, this one should a expected and an actual, right?

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+import javax.annotation.Nullable;
+
+/**
+ * The time trigger information to be passed to the triggered program.
+ */
+public class TimeTriggerInfo extends TriggerInfo {
+  private final String cronExpression;
+  @Nullable
+  private String logicalStartTime;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java]: Time shouldn't be a string.

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java]: and it cannot be null for time trigger, so should use `long`.

Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java]: This is null also because of the OR trigger, in which the TimeTrigger may not be satisfied

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java [hunk 11]
```diff
@@ -16,8 +16,79 @@
...
+   */
+  enum Type {
+    TIME("time"),
+    PARTITION("partition"),
+    STREAM_SIZE("stream-size"),
+    PROGRAM_STATUS("program-status"),
+    AND("and", true),
+    OR("or", true);
+
+    private static final Map<String, Type> CATEGORY_MAP;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java]: Why need this?? Just a plain enum is not enough?? It's only used programmatically to match the TriggerInfo type, right? Also, why not define it inside `TriggerInfo`?? Seems only used there.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java [hunk 12]
```diff
@@ -16,8 +16,79 @@
...
+      return type;
+    }
+
+    /**
+     * Whether the trigger type represents a composite trigger, i.e. a trigger that can
+     * contains multiple triggers internally.
+     *
+     * @return {@code true} if the trigger type represents a composite trigger, {@code false} otherwise
+     */
+    public boolean isComposite() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java]: How is it being used?? I don't see this method is used.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/TriggeringScheduleInfo.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  private final String name;
+  private final String description;
+  private final TriggerInfo triggerInfo;
+  private final Map<String, String> properties;
+
+  public TriggeringScheduleInfo(String name, String description, TriggerInfo triggerInfo,
+                                Map<String, String> properties) {
+    this.name = name;
+    this.description = description;
+    this.properties = properties;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/TriggeringScheduleInfo.java]: defensive copy the map.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 14]
```diff
@@ -212,6 +239,15 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
     }
   }
 
+  @Nullable
+  public static TriggeringScheduleInfo getTriggeringScheduleInfo(ProgramOptions programOptions) {
+    String scheduleInfoString = programOptions.getArguments().getOption(ProgramOptionConstants.TRIGGERING_SCHEDULE_INFO);
+    if (scheduleInfoString == null) {
+      return null;
+    }
+    return GSON.fromJson(scheduleInfoString, TriggeringScheduleInfo.class);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Don't decode it every time. Decode it once in the constructor.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Wait..I see it gets called in constructor, but why this is a public static method?? should be a private static one.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 15]
```diff
@@ -77,6 +77,12 @@
 
   public static final String USER_OVERRIDES = "userOverrides";
 
+  public static final String RUNTIME_ARGS = "runtimeArgs";
+
+  public static final String TRIGGERING_PROPERTIES_MAPPING = "triggering.properties.mapping";
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: Why it is defined in app-fabric??

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 16]
```diff
@@ -77,6 +77,12 @@
...
+
+  public static final String TRIGGERING_SCHEDULE_INFO = "triggeringScheduleInfo";
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: Same as above.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 17]
```diff
@@ -77,6 +77,12 @@
 
   public static final String USER_OVERRIDES = "userOverrides";
 
+  public static final String RUNTIME_ARGS = "runtimeArgs";
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: This is not used for getting things from system arguments. Shouldn't define in here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java (hunk 16 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: This is fine.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java [hunk 18]
```diff
@@ -97,6 +105,12 @@ public void launch(Job job) throws Exception {
     }
     systemArgs.put(ProgramOptionConstants.SCHEDULE_NAME, schedule.getName());
     systemArgs.put(ProgramOptionConstants.EVENT_NOTIFICATIONS, GSON.toJson(job.getNotifications()));
+    SatisfiableTrigger trigger = ((SatisfiableTrigger) job.getSchedule().getTrigger());
+    TriggerInfoContext triggerInfoContext = new TriggerInfoContext(job, store);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java]: So this is just for getting `TriggerInfo` below? Seems more like a private function in this class.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java]: Ideally in here you just create the `TriggeringScheduleInfo` (via a private method). That method can further divided, but it's not the concern in here (programming by intention).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java [hunk 19]
```diff
@@ -85,4 +88,14 @@ private void initializeUnitTriggers() {
       }
     }
   }
+
+  public List<TriggerInfo> getUnitTriggerInfos(TriggerInfoContext context) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java]: Line 37: field should be final.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java]: Or, if the field is not final, then the Map constructed Also, comment on line 36 is wrong, as a composite trigger can have another composite trigger inside, right?

Also, if you use `LinkedListMultimap` to store the triggers, then you can simply return an unmodifiable list of the multimap.values(), without the need of copying.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java [hunk 20]
```diff
@@ -45,11 +53,71 @@ public ProgramStatusTrigger(ProgramId programId, ProgramStatus... programStatuse
 
   @Override
   public boolean isSatisfied(List<Notification> notifications) {
-    return true;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java]: This change is not really related to adding TriggeringScheduleInfo, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java [hunk 21]
```diff
@@ -45,11 +53,71 @@ public ProgramStatusTrigger(ProgramId programId, ProgramStatus... programStatuse
...
+      }
+      ProgramStatus programStatus;
+      try {
+        programStatus = ProgramRunStatus.toProgramStatus(ProgramRunStatus.valueOf(programRunStatusString));
+      } catch (IllegalArgumentException e) {
+        // Return silently, this happens for statuses that are not meant to be scheduled
+        continue;
+      }
+      ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      ProgramId triggeringProgramId = programRunId.getParent();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java:130]: Seems like up to here this method is duplicating the code in `isSatisfied`. Better refactor them to a common private method.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 4 above)
Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: If it is unsatisfied, why it is included in the trigger lists?? It's not part of the upstream run that triggers the current program.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java (hunk 2 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java]: cdap-api module does not include guava as dependency, so just a copy into an ArrayList?

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 4 above)
Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: The `TriggeringScheduleInfo` should only contains triggers that actually satisfied (or if you really want to give all trigger info, then you need to have a boolean wasSatisfied method, although I don't think the list should include unsatisfied one).

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java (hunk 2 above)
Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/AbstractCompositeTriggerInfo.java]: See other class, but basically `Collections.unmodifiableList(new ArrayList<>(unitTriggerInfos))`

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java (hunk 11 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/Trigger.java]: It's used in `ProgramLifecycleHttpHandler`, following the similar style as the usage of `ProgramStatus` in REST API.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 4 above)
Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: By that, it also means the `TriggeringScheduleInfo` should simply just return a list of `TriggerInfo`, instead of having composite one.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java (hunk 20 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java]: no, it's just moving the logic to `ProgramStatusTrigger` from the subscriber service and handle the case of `ProgramStatusTrigger` in composite trigger

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java (hunk 19 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java]: yes, a composite trigger can contain other composite triggers, however, this method will only return the triggers which are not composite triggers, since composite triggers get trigger keys, trigger info and etc. from the internal unit triggers (i.e. non-composite triggers). I think LinkedHashMultimap will be better since it avoids duplicates in values?

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java]: huh?? That's quite confusing. I thought composite trigger is a tree of triggers, no??

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 4 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: that's true. I will just exclude unsatisfied ones

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java [hunk 22]
```diff
@@ -24,7 +24,7 @@
 public class TimeTriggerInfo extends TriggerInfo {
   private final String cronExpression;
   @Nullable
-  private String logicalStartTime;
+  private final String logicalStartTime;
```
Comment (sreevatsanraman) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/TimeTriggerInfo.java]: This should be long right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java (hunk 19 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java]: Yes, each composite trigger is like a node in the tree, and unit triggers are like the leaf nodes in the tree. The tree structure is not stored in the map, but is represented by the composite trigger which contains other triggers. This `getUnitTriggerInfos` method just need to return all the unique leaf nodes in the tree, i.e. unique unit triggers.

File: cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/SerializableSparkExecutionContext.scala [hunk 23]
```diff
@@ -128,4 +129,6 @@ class SerializableSparkExecutionContext(val delegate: SparkExecutionContext) ext
   }
 
   override def getDataTracer(loggerName: String): DataTracer = delegate.getDataTracer(loggerName)
+
+  override def getTriggeringScheduleInfo: TriggeringScheduleInfo = delegate.getTriggeringScheduleInfo
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/SerializableSparkExecutionContext.scala]: It is not common to have `null` value in Scala in general (which the `getTriggeringScheduleInfo` might return null).

One possible way to get around that is to have a separate interface to have the `getTriggeringScheduleInfo` method (e.g. `SchedulableProgramContext`), which program context, except the SparkExecutionContext, implements. The `SparkExecutionContext` can have it's own definition of `getTriggeringScheduleInfo`, which returns an `Option[TriggeringScheduleInfo]`.

File: cdap-api/src/main/java/co/cask/cdap/api/SchedulableRuntimeContext.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+package co.cask.cdap.api;
+
+import co.cask.cdap.api.schedule.TriggeringScheduleInfo;
+
+import javax.annotation.Nullable;
+
+/**
+ * This interface represents a context for a processor or elements of a processor.
+ */
+public interface SchedulableRuntimeContext extends RuntimeContext {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/SchedulableRuntimeContext.java]: Don't extends from `RuntimeContext`, and better call this interface `SchedulableProgramContext`.

File: cdap-api-spark-base/src/main/scala/co/cask/cdap/api/spark/SparkExecutionContextBase.scala [hunk 25]
```diff
@@ -292,4 +292,12 @@ trait SparkExecutionContextBase extends RuntimeContext with Transactional {
     */
   @throws(classOf[IOException])
   def createInterpreter(): SparkInterpreter
+
+  /**
+    * Get the information of the schedule that launches this Spark program, if there is any.
+    *
+    * @return a instance of [[Some]] containing [[TriggeringScheduleInfo]] of the schedule that launches this
```
Comment (chtyim) [inline cdap-api-spark-base/src/main/scala/co/cask/cdap/api/spark/SparkExecutionContextBase.scala]: Scala doc link need to use full qualified class name (e.g. `[[scala.Some]]`).

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+ * The program status trigger information to be passed to the triggered program.
+ */
+public class ProgramStatusTriggerInfo extends TriggerInfo {
+
+  private final String namespace;
+  private final ApplicationSpecification applicationSpecification;
+  private final ProgramType programType;
+  private final String program;
+  private final Set<ProgramStatus> triggerStatues;
+  private final String runId;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: runId should be of type `RunId` (from `org.apache.twill.api.RunId`).

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+   * @return The program name of the triggering program.
+   */
+  public String getProgram() {
+    return program;
+  }
+
+  /**
+   * @return An immutable set of all the program statuses that can satisfy the program status trigger.
+   */
+  public Set<ProgramStatus> getTriggerStatues() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: What is this for?? The triggering info should only contains information of what actually happened. I think we only need the `getProgramStatus` below.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/PartitionTriggerInfo.java [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.schedule;
+
+/**
+ * The dataset partition trigger information to be passed to the triggered program.
+ */
+public class PartitionTriggerInfo extends TriggerInfo {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/PartitionTriggerInfo.java]: Each implementation of `TriggerInfo` should be serializable in order to be useable in Spark. You might need to implements the read/write Object methods or implementing the `Externalizable` directly.

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/PartitionTriggerInfo.java]: E.g. ApplicationSpecification, RunId are both interfaces, hence you need custom read/write logic for them. Also, you might want to define the `TriggerInfo` hierarchy as interfaces in order to be able to do serialization with non-api classes.

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+package co.cask.cdap.api.schedule;
+
+/**
+ * The stream size trigger information to be passed to the triggered program.
+ */
+public class StreamSizeTriggerInfo extends TriggerInfo {
+  private final String streamNamespace;
+  private final String streamName;
+  private final int triggerMB;
+  private final long logicalStartTime;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java]: Why it has logical start time for stream size trigger?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java [hunk 30]
```diff
@@ -77,31 +85,31 @@ public void launch(Job job) throws Exception {
     ProgramSchedule schedule = job.getSchedule();
     ProgramId programId = schedule.getProgramId();
     Map<String, String> userArgs = Maps.newHashMap();
-    Map<String, String> systemArgs = Maps.newHashMap();
-    // notificationProperties is present only in jobs containing schedules with TimeTrigger and StreamSizeTrigger.
-    // Since both triggers are satisfied by the first notification, there can be only one notification in in the job
-    Map<String, String> notificationProperties = job.getNotifications().get(0).getProperties();
     userArgs.putAll(schedule.getProperties());
     userArgs.putAll(propertiesResolver.getUserProperties(programId.toId()));
-    String userOverridesString = notificationProperties.get(ProgramOptionConstants.USER_OVERRIDES);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java:86]: What are user overrides and system overrides ??

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java (hunk 26 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/ProgramStatusTriggerInfo.java]: Since only `ProgramId` is available in notifications, should I call `RunIds.fromString(programRunId.getRun()` to convert the runId String to RunId?

File: cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java (hunk 29 above)
Comment (ecfm) [reply cdap-api/src/main/java/co/cask/cdap/api/schedule/StreamSizeTriggerInfo.java]: that's the time when the stream size reaches the threshold to satisfy the trigger

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java (hunk 30 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java:86]: they are the runtime arguments that schedules want to add.

File: cdap-unit-test/src/test/scala/co/cask/cdap/spark/app/TriggeredSpark.scala [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+    setMainClass(classOf[TriggeredSpark])
+  }
+
+  override def run(implicit sec: SparkExecutionContext): Unit = {
+    val scheduleInfoOption: Option[TriggeringScheduleInfo] = sec.getTriggeringScheduleInfo
+    Preconditions.checkState(scheduleInfoOption.isDefined)
+    val scheduleInfo: TriggeringScheduleInfo = scheduleInfoOption.get
+    val triggerInfo: ProgramStatusTriggerInfo =
+      scheduleInfo.getTriggerInfos.get(0).asInstanceOf[ProgramStatusTriggerInfo]
+    Preconditions.checkState(classOf[TestSparkApp].getSimpleName
```
Comment (chtyim) [inline cdap-unit-test/src/test/scala/co/cask/cdap/spark/app/TriggeredSpark.scala]: You have to run something in the executor to verify that the serialization is actually working. E.g.

```
val triggerInfo = sec.getTriggeringScheduleInfo.get.getTriggerInfos.get(0).asInstanceOf[ProgramStatusTriggerInfo]
val sc = new SparkContext
val sum = sc.parallelize(Seq(1, 2, 3, 4, 5)).filter(triggerInfo.getProgram().equals("ForkSparkWorkflow")).reduce(_ + _)
require(sum == 15)
```

Comment (ecfm) [reply cdap-unit-test/src/test/scala/co/cask/cdap/spark/app/TriggeredSpark.scala]: addressed and fixed deserialization

**Comment by chtyim** [comment]
Updated the PR with 8 commits, each commit message should tell what the changes are.

Didn't expect this PR has data-pipeline changes as well, which I am not familiar with, better to have @albertshau to take a look.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java [hunk 32]
```diff
@@ -57,6 +58,21 @@ public DefaultMacroEvaluator(@Nullable WorkflowToken workflowToken, Map<String,
     this.namespace = namespace;
   }
 
+  public DefaultMacroEvaluator(BasicArguments arguments,
+                               long logicalStartTime, SecureStore secureStore, String namespace) {
+    Map<String, String> resolvedArguments = new HashMap<>();
+    Iterator<Map.Entry<String, String>> iter = arguments.iterator();
+    while (iter.hasNext()) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java:65]: Use for-each instead of `Iterator` since BasicArguments implements `Iterable`.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java [hunk 33]
```diff
@@ -57,6 +58,21 @@ public DefaultMacroEvaluator(@Nullable WorkflowToken workflowToken, Map<String,
     this.namespace = namespace;
   }
 
+  public DefaultMacroEvaluator(BasicArguments arguments,
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultMacroEvaluator.java:61]: should remove the other constructor and replace all usages of it with this one

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 34]
```diff
@@ -254,11 +270,103 @@ protected void configure() {
...
+    if (value != null) {
+      // Put the triggering pipeline property value to the runtime argument with the key in current pipeline
+      runtimeArgs.put(currentKey, value);
+    }
+  }
+
   @Override
   public void initialize(WorkflowContext context) throws Exception {
     super.initialize(context);
+    Map<String, String> combinedRuntimeArgs = new HashMap<>(context.getRuntimeArguments());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: no need for this. after you add things to the token, create the PipelineRuntime, and use pipelineRuntime.getArguments().asMap() instead.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 35]
```diff
@@ -254,11 +270,103 @@ protected void configure() {
     addPrograms(start, programAdder);
   }
 
+  private Map<String, String> getNewRuntimeArgsFromScheduleInfo(TriggeringScheduleInfo scheduleInfo,
+                                                                Map<String, String>
+                                                                  propertiesMap) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 36]
```diff
@@ -254,11 +270,103 @@ protected void configure() {
...
+    List<TriggerInfo> triggerInfoList = scheduleInfo.getTriggerInfos();
+    List<ProgramStatusTriggerInfo> programStatusTriggerInfos = new ArrayList<>();
+    for (TriggerInfo info : triggerInfoList) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: just do `TriggerInfo info : scheduleInfo.getTriggerInfos()`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 37]
```diff
@@ -254,11 +270,103 @@ protected void configure() {
...
+      if (info instanceof ProgramStatusTriggerInfo) {
+        programStatusTriggerInfos.add((ProgramStatusTriggerInfo) info);
+      }
+    }
+    Map<String, String> newRuntimeArgs = new HashMap<>();
+    // If no ProgramStatusTriggerInfo, no need of override the existing runtimeArgs
+    if (programStatusTriggerInfos.size() == 0) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: when checking if a collection is empty, use isEmpty() instead of comparing size to 0

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 38]
```diff
@@ -254,11 +270,103 @@ protected void configure() {
...
+      return newRuntimeArgs;
+    }
+    // Get the value of every triggering pipeline property and put it to newRuntimeArgs if it's not null
+    for (Map.Entry<String, String> entry : propertiesMap.entrySet()) {
+      TriggeringPipelinePropertyId triggeringPropertyId =
+        GSON.fromJson(entry.getKey(), TriggeringPipelinePropertyId.class);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: there is no need for this. When you originally deserialize it, you should deserialize it to a `Map<String, TriggeringPipelinePropertyId>`

**Comment by ecfm** [comment]
Split this PR into platform changes PR and data pipeline app changes PR
